modules/src/core/ics04_channel/context.rs

```diff
@@ -280,6 +272,71 @@ pub trait ChannelKeeper {
     fn increase_channel_counter(&mut self);
 }
 
+pub trait ChannelCapabilityReader<M: Into<ModuleId>>: CapabilityReader<M> {
+    /// Return the `ModuleId` along with the `Capability` associated with a given
+    /// `PortId`-`ChannelId`
+    fn lookup_module_by_channel(
+        &self,
+        channel_id: ChannelId,
+        port_id: PortId,
+    ) -> Result<(ModuleId, Self::Capability), Error> {
+        self.lookup_module(&channel_capability_name(port_id, channel_id))
+            .map_err(Error::ics05_port)
+    }
+
+    /// Get the `Capability` associated with the specified `PortId`-`ChannelId`
+    fn get_channel_capability(
+        &self,
+        port_id: PortId,
+        channel_id: ChannelId,
+    ) -> Result<Self::Capability, Error> {
+        self.get_capability(&channel_capability_name(port_id, channel_id))
+            .map_err(Error::ics05_port)
+    }
+
+    /// Authenticate a `Capability` against the specified `PortId`-`ChannelId` by checking if
+    /// the capability was previously generated and bound to the specified port/channel
+    fn authenticate_channel_capability(
+        &self,
+        port_id: PortId,
+        channel_id: ChannelId,
+        capability: &Self::Capability,
+    ) -> Result<(), Error> {
+        self.authenticate_capability(&channel_capability_name(port_id, channel_id), capability)
+            .map_err(Error::ics05_port)
+    }
+
+    fn create_channel_capability(
```

Comment:
This is an important deviation from the SDK's capability design. We require implementations to be able to create capabilities without mutating the state. The capability returned from this call used then stored to the state by calling `ChannelCapabilityKeeper::store_channel_capability()`. This is required because our handlers do not mutate state, but still need to be able to create capabilities.